### DISCUSSION TITLE
undefined reference to `Arduino_ESP32SPI::Arduino_ESP32SPI

### DISCUSSION THREAD

**Discussion by MrDomingo** [moononournation/Arduino_GFX#77] [Q&A] [answered]
Hi,
I just acquired a 4" display based on ILI9486 and I am trying to test it using a ESP32 dev module.  The module seems to be a copy of Do It ESP32 DevKit.  I attempted to build "HelloWorld" but I'm getting error: undefined reference to `Arduino_ESP32SPI::Arduino_ESP32SPI(...)'
In function `_GLOBAL__sub_I_bus'

To get even here I had to add a line
#define CONFIG_IDF_TARGET_ESP32 true
before the #include statement.  I also added following in the #else block:
#define TFT_CS  5
#define TFT_DC 27
#define TFT_RST 33
#define TFT_BL 22

Arduino_DataBus *bus = new Arduino_ESP32SPI(TFT_DC,TFT_CS, 18 /* SCK */, 23 /* MOSI */,-1 /* MISO */, VSPI /* spi_num */);
// ILI9486 LCD 320x480
Arduino_GFX *gfx = new Arduino_ILI9486_18bit(bus, TFT_RST, 0 /* rotation */, false /* IPS */);

Build fails with the error above.  I checked what was being built in C:\Users\<user>\AppData\Local\Temp\arduino_build_111497 and all seems fine.  all of your library is being build and .o files were generated and were part of the link command.

I tried switching to 1.1.4 version of your library but it made no difference.

any ideas?

**Comment by MrDomingo** [comment, marked as answer]
Resolved it.  Switched to Platform IO and that allowed me to investigate more effectively.
Basically I needed to follow guide more closely.  This worked for me:

#define CONFIG_IDF_TARGET_ESP32 true

#define TFT_CS 5
#define TFT_DC 27
#define TFT_RST 33
#define TFT_BL 22

#include <Arduino_GFX_Library.h>

/* More dev device declaration: https://github.com/moononournation/Arduino_GFX/wiki/Dev-Device-Declaration */
#if defined(DISPLAY_DEV_KIT)
Arduino_GFX *gfx = create_default_Arduino_GFX();
#else /* !defined(DISPLAY_DEV_KIT) */

/* More data bus class: https://github.com/moononournation/Arduino_GFX/wiki/Data-Bus-Class */
Arduino_DataBus *bus = create_default_Arduino_DataBus();

/* More display class: https://github.com/moononournation/Arduino_GFX/wiki/Display-Class */
///Arduino_GFX *gfx = new Arduino_ILI9341(bus, TFT_RST, 0 /* rotation */, false /* IPS */);

// ILI9486 LCD 320x480
Arduino_GFX *gfx = new Arduino_ILI9486_18bit(bus, TFT_RST, 0 /* rotation */, false /* IPS */);

#endif /* !defined(DISPLAY_DEV_KIT) */

**Reply by moononournation** [reply]
But I don't know why you need explicitly #define CONFIG_IDF_TARGET_ESP32 true

**Reply by MrDomingo** [reply]
I did not see any evidence that it was defined anywhere else and if not defined PlatformIO GUI indicated that in function below it would be compiling the default block on the bottom.  I did not know whether THAT default function would somehow work due to some other library default behavior.

Arduino_DataBus *create_default_Arduino_DataBus()
{
#if defined(ARDUINO_ARCH_NRF52840)
    return new Arduino_NRFXSPI(TFT_DC, TFT_CS, TFT_SCK, TFT_MOSI, TFT_MISO);
#elif defined(ARDUINO_RASPBERRY_PI_PICO)
    return new Arduino_RPiPicoSPI(TFT_DC, TFT_CS, TFT_SCK, TFT_MOSI, TFT_MISO, spi0);
#elif CONFIG_IDF_TARGET_ESP32 // ESP32/PICO-D4
    return new Arduino_ESP32SPI(TFT_DC, TFT_CS, TFT_SCK, TFT_MOSI, TFT_MISO);
#elif defined(ESP8266)
    return new Arduino_ESP8266SPI(TFT_DC, TFT_CS);
#else
    return new Arduino_HWSPI(TFT_DC, TFT_CS);
#endif
}

**Reply by moononournation** [reply]
it is already defined in esp-idf

**Reply by MrDomingo** [reply]
Not according to PlatformIO.  When I comment out my declaration, PlatformIO cannot find any other declaration.  Supposedly it should be in "sdkconfig.h" but I did not find it in my (Arduino) sdkconfig.h file and I do not know where PlatformIO keeps its copies of Espressif source code.  It could be defined in some make file or some associated file.

**Reply by moononournation** [reply]
just tested Arduino IDE does not have this problems
